### PR TITLE
A slightly manual way of recording history.

### PR DESCRIPTION
**Pull request by ashimali**
To record history for a given class, you'd need to add the versioned mixin in models.py 
to the class you want to record history for.

Then create history model to mirror the versioned class with fields you want to record from the versioned class.

You can then use versioned decorator from dao_utils on the dao methods where your class is created or updated.

The history table records current state on create or update. This removes need to try and diff changes.

A consequence of this means that versioned objects should be given their id cols manually. By doing this the history models can be saved in the same transaction and rolled back with the versioned object in case of errors.

### CODE REVIEW

File: app/models.py [hunk 1]
```diff
@@ -64,14 +73,30 @@ def check_password(self, password):
...
     db.Column('user_id', UUID(as_uuid=True), db.ForeignKey('users.id')),
     db.Column('service_id', UUID(as_uuid=True), db.ForeignKey('services.id')),
-
     UniqueConstraint('user_id', 'service_id', name='uix_user_to_service')
 
 )
 
 
-class Service(db.Model):
+class ServiceHistory(db.Model):
```
Comment (quis) [inline app/models.py]: Why can’t this inherit from the `Service` model and override the things that are different?

File: app/service/rest.py [hunk 2]
```diff
@@ -1,3 +1,4 @@
+import uuid
```
Comment (quis) [inline app/service/rest.py]: Why is this import being added?

File: app/service/rest.py [hunk 3]
```diff
@@ -72,6 +73,9 @@ def create_service():
 
     data.pop('user_id', None)
 
+    # import pdb
```
Comment (quis) [inline app/service/rest.py]: Why is this here?

File: app/vendor/sqlalchemy_history_meta.py [hunk 4]
```diff
@@ -0,0 +1,261 @@
+"""
```
Comment (mannickutd) [inline app/vendor/sqlalchemy_history_meta.py]: Is this file still needed?

Comment (quis) [inline app/vendor/sqlalchemy_history_meta.py]: Are you still using this file?

**Comment by quis** [comment]
I’ve left some comments on the code. Someone else should take a look at it as well, but I think the concept is good.

File: migrations/versions/0002_add_service_history.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
+"""empty message
```
Comment (servingUpAces) [inline migrations/versions/0002_add_service_history.py]: This upgrade does not work on a populated db.

File: app/dao/dao_utils.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
+from functools import wraps
+
```
Comment (mannickutd) [inline app/dao/dao_utils.py]: I do like these decorators.
One scenario which we may have to alter their functionality in the future is to handle situations where we want to commit and situations where it is one of a many db operations and only on the last db operation do we want to commit. How do we handle the stacked decorated methods?

File: app/models.py (hunk 1 above)
Comment (ashimali) [inline app/models.py]: Most of the fields are nullable, the main difference is the composite primary key of id and version, so I think inheritance wouldn't be the way to go for this model.

File: app/service/rest.py (hunk 2 above)
Comment (ashimali) [inline app/service/rest.py]: stray import

File: app/service/rest.py (hunk 3 above)
Comment (ashimali) [inline app/service/rest.py]: removed

File: app/vendor/sqlalchemy_history_meta.py (hunk 4 above)
Comment (ashimali) [inline app/vendor/sqlalchemy_history_meta.py]: No will remove it.

File: migrations/versions/0002_add_service_history.py (hunk 5 above)
Comment (ashimali) [inline migrations/versions/0002_add_service_history.py]: Good point. Will do an update to set existing models version to 1 and also create a history record for each.

File: app/dao/dao_utils.py (hunk 6 above)
Comment (ashimali) [inline app/dao/dao_utils.py]: Yep there is a required order for the decorators.

File: tests/app/service/test_rest.py [hunk 7]
```diff
@@ -212,15 +213,16 @@ def test_create_service(notify_api, sample_user):
...
     with notify_api.test_request_context():
         with notify_api.test_client() as client:
             data = {
                 'email_from': 'service',
                 'name': 'created service',
                 'message_limit': 1000,
                 'restricted': False,
-                'active': False
+                'active': False,
+                'created_by': str(fake_uuid)
```
Comment (mannickutd) [inline tests/app/service/test_rest.py]: Don't need to str the fake_uuid but doesn't matter for now.

**Comment by mannickutd** [comment]
:+1:
